File: benchmarks/livermore_canonical/k16_monte_carlo.f
C     Kernel 16 -- Monte Carlo search loop (canonical)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C     Canonical body: lines 351-385 of the kernels listing.
C     R, S, T pulled from canonical SPACER positions 30, 32, 36.
C
      PROGRAM K16
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP, ZONE
      DIMENSION D(300), PLAN(300), ZONE(300)
      DIMENSION SPACER(39)
C
C     SPACER -> R, S, T at positions 30, 32, 36
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 5 K = 1, 39
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         SPACER(K) = (BUZZ - FIZZ) * 0.1D0
    5 CONTINUE
      R = SPACER(30)
      S = SPACER(32)
      T = SPACER(36)
C
C     PLAN, D via SIGNEL (independent streams)
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 10 I = 1, 300
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         PLAN(I) = (BUZZ - FIZZ) * 0.1D0
   10 CONTINUE
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 12 I = 1, 300
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         D(I) = (BUZZ - FIZZ) * 0.1D0
   12 CONTINUE
C
C     ZONE(i) = (i-1) mod (n+1); ZONE(1) = 5
C
      N  = 75
      II = N / 3
      LB = II + II
      DO 15 I = 1, 300
         ZONE(I) = MOD(I-1, N+1)
   15 CONTINUE
      ZONE(1) = 5
C
C     Kernel 16 body (canonical state machine, lines 357-385)
C
      DO 500 REP = 1, 162000000
         M  = 1
         I1 = M
         K2 = 0
         K3 = 0
  410    J2 = (N+N)*(M-1) + 1
         DO 470 K = 1, N
            K2 = K2 + 1
            J4 = J2 + K + K
            J5 = ZONE(J4)
            IF (J5 .LT. N) GO TO 420
            IF (J5 .EQ. N) GO TO 475
            GO TO 450
  415       IF (J5 .LT. N - II) GO TO 430
            GO TO 425
  420       IF (J5 .LT. N - LB) GO TO 435
            GO TO 415
  425       TMP = PLAN(J5) - R
            GO TO 446
  430       TMP = PLAN(J5) - S
            GO TO 446
  435       TMP = PLAN(J5) - T
            GO TO 446
  440       IF (ZONE(J4-1) .LT. 0) GO TO 455
            IF (ZONE(J4-1) .EQ. 0) GO TO 485
            GO TO 470
  445       IF (ZONE(J4-1) .LT. 0) GO TO 470
            IF (ZONE(J4-1) .EQ. 0) GO TO 485
            GO TO 455
  446       IF (TMP .LT. 0.0D0) GO TO 445
            IF (TMP .EQ. 0.0D0) GO TO 480
            GO TO 440
  450       K3 = K3 + 1
            DEXPR = D(J5) - (D(J5-1)*(T-D(J5-2))**2
     &                       + (S-D(J5-3))**2
     &                       + (R-D(J5-4))**2)
            IF (DEXPR .LT. 0.0D0) GO TO 445
            IF (DEXPR .EQ. 0.0D0) GO TO 480
            GO TO 440
  455       M = M + 1
            IF (M .GT. ZONE(1)) GO TO 460
            GO TO 465
  460       M = 1
  465       IF (I1 .EQ. M) GO TO 480
            GO TO 410
  470    CONTINUE
  475    CONTINUE
  480    CONTINUE
  485    CONTINUE
  500 CONTINUE
C
      WRITE(*,910) K2
      WRITE(*,920) K3
  910 FORMAT('k2 = ', I20)
  920 FORMAT('k3 = ', I20)
      END

File: benchmarks/livermore_canonical/k16_monte_carlo.c
/* K16 -- Monte Carlo search loop (canonical, 1-based indexing)
   Canonical body: lines 351-385 of the kernels listing.
   R, S, T from canonical SPACER positions 30, 32, 36. */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     75
#define NREPS 162000000

int main(void) {
    static double d[301], plan[301];
    static long zone[301];
    static double spacer[40];

    /* SPACER -> R, S, T */
    double fuzz = 1.2345e-3, buzz = 1.0 + fuzz, fizz = 1.1 * fuzz;
    for (long k = 1; k <= 39; k++) {
        buzz = (1.0 - fuzz) * buzz + fuzz;
        fuzz = -fuzz;
        spacer[k] = (buzz - fizz) * 0.1;
    }
    double r = spacer[30], s = spacer[32], t = spacer[36];

    /* PLAN, D via SIGNEL (independent streams) */
    fuzz = 1.2345e-3; buzz = 1.0 + fuzz; fizz = 1.1 * fuzz;
    for (long i = 1; i <= 300; i++) {
        buzz = (1.0 - fuzz) * buzz + fuzz;
        fuzz = -fuzz;
        plan[i] = (buzz - fizz) * 0.1;
    }
    fuzz = 1.2345e-3; buzz = 1.0 + fuzz; fizz = 1.1 * fuzz;
    for (long i = 1; i <= 300; i++) {
        buzz = (1.0 - fuzz) * buzz + fuzz;
        fuzz = -fuzz;
        d[i] = (buzz - fizz) * 0.1;
    }

    /* ZONE */
    long ii = N / 3, lb = ii + ii;
    for (long i = 1; i <= 300; i++)
        zone[i] = (i - 1) % (N + 1);
    zone[1] = 5;

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    long k2 = 0, k3 = 0;
    long m, i1, j2, j4, j5;
    double tmp, dexpr;

    for (long rep = 1; rep <= NREPS; rep++) {
        m = 1;
        i1 = m;
        k2 = 0;
        k3 = 0;
    L410:
        j2 = (N + N) * (m - 1) + 1;
        for (long k = 1; k <= N; k++) {
            k2 = k2 + 1;
            j4 = j2 + k + k;
            j5 = zone[j4];
            if (j5 < N) goto L420;
            if (j5 == N) goto L475;
            goto L450;
        L415:
            if (j5 < N - ii) goto L430;
            goto L425;
        L420:
            if (j5 < N - lb) goto L435;
            goto L415;
        L425:
            tmp = plan[j5] - r;
            goto L446;
        L430:
            tmp = plan[j5] - s;
            goto L446;
        L435:
            tmp = plan[j5] - t;
            goto L446;
        L440:
            if (zone[j4-1] < 0) goto L455;
            if (zone[j4-1] == 0) goto L485;
            continue;  /* GO TO 470 */
        L445:
            if (zone[j4-1] < 0) continue;  /* GO TO 470 */
            if (zone[j4-1] == 0) goto L485;
            goto L455;
        L446:
            if (tmp < 0.0) goto L445;
            if (tmp == 0.0) goto L480;
            goto L440;
        L450:
            k3 = k3 + 1;
            dexpr = d[j5] - (d[j5-1]*(t-d[j5-2])*(t-d[j5-2])
                             + (s-d[j5-3])*(s-d[j5-3])
                             + (r-d[j5-4])*(r-d[j5-4]));
            if (dexpr < 0.0) goto L445;
            if (dexpr == 0.0) goto L480;
            goto L440;
        L455:
            m = m + 1;
            if (m > zone[1]) m = 1;
            if (i1 == m) goto L480;
            goto L410;
        }
    L475: ;
    L480: ;
    L485: ;
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("k2 = %ld\n", k2);
    printf("k3 = %ld\n", k3);
    return 0;
}
